src/Microsoft.OData.Core/UriParser/SemanticAst/SelectExpandClauseExtensions.cs

```diff
@@ -83,7 +83,20 @@ internal static void GetSelectExpandPaths(this SelectExpandClause selectExpandCl
         /// <returns>String list generated from selected items</returns>
         internal static List<string> GetCurrentLevelSelectList(this SelectExpandClause selectExpandClause)
         {
-            return selectExpandClause.SelectedItems.Select(GetSelectString).Where(i => i != null).ToList();
+            if (selectExpandClause.AllAutoSelected)
+            {
+                return selectExpandClause.SelectedItems
+                    .OfType<ExpandedNavigationSelectItem>()
+                    .Select(i => String.Join("/", i.PathToNavigationProperty.WalkWith(PathSegmentToStringTranslator.Instance).ToArray()))

```

Comment:
Please notice that there is a static method  `String.Join` that accepts `IEnumerable<T>`. If you apply it here you will avoid extra memory allocations.

https://msdn.microsoft.com/en-us/library/dd992421(v=vs.110).aspx